File: linked-list/06-get-node-value.ts
// import classes and types
import { dummy_llist } from "./utils/mock";
import { LinkedListNode } from "./utils/types";

// Function : to get node value from any position
const get_node_value = (head: null | LinkedListNode, pos: number): null | number => {
    if (head === null || pos < 0) {
        return null;
    };
    let curr_node = head;
    let curr_position = 0;
    while (curr_position !== pos && curr_node) {
        curr_position++;
        curr_node = curr_node.next!;
    };
    return curr_node?.data || null;
};

// print initial value of llist
console.log("INITIAL LLIST");
dummy_llist.print();

// get nodes data from random positions
console.log("GETTING DATA");
console.log(`Data at index (0) is: `, get_node_value(dummy_llist.head, 0));
console.log(`Data at index (3) is: `, get_node_value(dummy_llist.head, 3));
console.log(`Data at index (4) is: `, get_node_value(dummy_llist.head, 4));
// testing : position = -1 (underflow)
console.log(`Data at index (-1) is: `, get_node_value(dummy_llist.head, -1));
// testing : position = 500 (overflow)
console.log(`Data at index (500) is: `, get_node_value(dummy_llist.head, 500));

// OUTPUT
// INITIAL LLIST
// {"data":1,"next":{"data":2,"next":{"data":3,"next":{"data":4,"next":{"data":5,"next":null}}}}}
// GETTING DATA
// Data at index (0) is:  1
// Data at index (3) is:  4
// Data at index (4) is:  5
// Data at index (-1) is:  null
// Data at index (500) is:  null

File: linked-list/06-get-node-value.py
# ==============================================================================================

class LinkedList():  # class LinkedList
    def __init__(self):
        self.head = None

    # method to insert node
    def insert_node(self, data: int):
        if(self.head == None):
            self.head = Node(data)
            return
        curr_node = self.head
        while(curr_node.next):
            curr_node = curr_node.next
        curr_node.next = Node(data)

    # method to print node
    def print_list(self):
        print('The Linked List', end=" ")
        curr_node = self.head
        while(curr_node):
            print("->", curr_node.data, end="")
            curr_node = curr_node.next
        print()


class Node:  # class Node
    def __init__(self, data):
        self.data = data
        self.next = None


# Function: to get node at any position
def get_node_at_pos(head: None | Node, position: int) -> None | int:
    if(head == None or position < 0):
        return None
    curr_node = head
    curr_position = 0
    while(curr_position != position and curr_node):
        curr_position += 1
        curr_node = curr_node.next
    if(curr_node is None):
        return None
    return curr_node.data


# create a linked list
my_llist = LinkedList()

# inserting elements to fill the list
my_llist.insert_node(10)
my_llist.insert_node(20)
my_llist.insert_node(30)
my_llist.insert_node(40)
my_llist.insert_node(50)

my_llist.print_list()

print()

# get elements at any position
print('node at position 3 is    :', get_node_at_pos(my_llist.head, 3))
print('node at position 0 is    :', get_node_at_pos(my_llist.head, 0))
print('node at position 1 is    :', get_node_at_pos(my_llist.head, 1))
# testing: position = -100 (underflow)
print('node at position -100 is :', get_node_at_pos(my_llist.head, -100))
# testing: position = 100 (overflow)
print('node at position 100 is  :', get_node_at_pos(my_llist.head, 100))

# OUTPUT
# The Linked List -> 10-> 20-> 30-> 40-> 50

# node at position 3 is    : 40
# node at position 0 is    : 10
# node at position 1 is    : 20
# node at position -100 is : None
# node at position 100 is  : None
